### ISSUE TITLE
Prepare Data for Analysis

### ISSUE THREAD

**Issue by cqcampos** [cqcampos/lausdinfoRCT#4]
This is the first task for this project. I describe the raw datasets you'll be working with below. You should not make any edits to these data and instead create an analysis dataset. All of these data can be found on Mercury at /project/lausd/rawdata/Campos_Master_DUA. Note that you will need to unzip files. 

**Some preliminaries about the data:** 
1. The studentidentifier across all datasets is studentpseudoid (or a variant of that). This is the identifier that links students across years, schools, semesters, etc 
2. There is a subtlety about schools/campuses and "programs". Some schools house multiple programs in a campus. Some of these programs are considered schools and others are not. Preferredlocationcode is the identifier for the campus and schoollocationcode is the identifier for the program. For many schools/programs, you will see that preferredlocationcode==schoollocationcode does not. I make a point out of this because magnet programs, the programs which will be important for this analysis, are usually defined by their schoollocationcode. We can return to this when it comes up but this is an important detail to keep in mind. 
3. The different datasets are mostly at the student level, with different spreadsheets corresponding to different years. Not all datasets are at the student level, so the first thing you should do when working with the raw data is confirm the unit of observation. 

**Goal**
Create a student-by-year dataset containing information about students' demographics, addresses, enrollment, interaction with the choice system, test scores, behavioral incidents, and GPA. 

**Description of Raw Data**
1. Demographics: A folder called 'Demographics' contains files named Demographics_Gr_01-12_YEAR.csv for YEAR={2002, 2003, ..., 2023}. The year corresponds to the second year of a given academic year (e.g., Demographics/Demographics_Gr_01-12_2009.csv corresponds to demographic data for the 2008-2009 academic year). We should consider this as the base data when creating the panel. In other words, we'll do many "left-joins" with the panel of demographics data being the "left" dataset. In the cleaning, create ethnicity/race indicators and indicators for whatever other relevant groups are defined in these data. 
2. Addresses: This folder contains student addresses that must be geocoded. I'm going to have Ryan do this and he'll share with you once ready. For your purposes, you will receive a panel (student-by-year) containing information about students residential census block in each academic year. 
3. Enrollment: Enrollment is contained in the demographics data discussed above. The preferredlocationcode/preferredlocationname and schoollocationcode/schoollocationname variables indicate the school the student was enrolled at at the end of the academic year
4. Interaction with the choice system: The folder 'Unified_Enrollment' contains information about students' interactions with the choice system in Los Angeles. Some preliminaries about identifiers here first. The various datasets in this folder contain student_pseudo_id (which you can rename studentpseudoid), application_id (identifier for the application), and school_yr (it's unclear what the convention for this school_yr is so figure that out when you work on this). You will notice that there are duplicates within student_pseudo_id-school_yr but not within application_id-school_yr, and that is because some students apply multiple times so they will have more than a single application. Be mindful of this when working to create the student-by-year panel. 
5. Test scores - For now, there is a folder called SBA (meaning Smarter Balanced Assessment) that contains test score data from 2015-2023. The unit of observation is studentpseudoid-schoolyear-testgroupname. Testgroupname corresponds to either ELA (English and Language Arts) or Math. Be mindful of this when creating the student-by-year panel. overallscalescore and overallperformancelevelname are the primary variables to keep.
6. Behavioral incidents: There is a folder called 'Suspensions' that you will work with. The data are at the incident level, meaning that for every suspension incident there is a row. The associated information is the studentpseudoid, the suspension date, the suspension type, and the number of suspended days. We want to create a student-by-year panel tells us the number of suspension incidents a student had in a given year, the total number of suspended days, the total number of class, in-school, and out-of-school suspension incidents, and for each category, the total number of suspended days. 

That's what comes to mind about the data above but I know you will have questions. Let's ask them here and make sure to tag me so I get a notification. Issues here are also linked to Asana so you can continue the conversation there. The advantage of communicating here is that if we ever need to write math, we can do that here. 

One final point: I would approach this build in the following way. First, for each dataset or type of datasets, determine what the key variables are, any medications that need to be done to them, and then determine what needs to be done so that you can create a student-by-year panel. This will be straightforward for most datasets but not all. Second, once you have a student-by-year panel for each subset of data, you can merge them all together. Once that merge is done, we should be done.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1206235607391825/1206265993553789) by [Unito](https://www.unito.io)

**Comment by cqcampos** [comment]
I also have to share a dataset that determines who was in the treatment group and who wasn't. I'll add that to Mercury later today. For now, you can work on the build as described above.

**Comment by aparthan** [comment]
@cqcampos I have been working with the datasets in folders Demographics, SBAC, Suspensions, and Unified_Enrollment individually to reshape wherever needed and create studentpseudoid-year data. I am then appending all years so I get a demographics_all, sbac_all datasets , etc. which can then be merged together. I have : 

1. Appended all the SBAC Performance files into a single data file titled SBAC_all. I reshaped the data to be at the studentpseudoid-endyear level with variables score_ELA, score_Math, etc. 
2. Written code to append demographic files (have not been able to run fully on terminal yet) 
3. Gone through the related data files for this build and identified some questions we should address highlighted below

Some things I want to flag about the data: 
1. In SBA folder we have files for 2015, 2016, 2017, 2018, 2019, 2022 and 2023. So we do not have scores for endyears 2020 and 2021 (2019-2020, 2020 - 2021 academic years respectively). I imagine this might be a covid thing and the specific tests were not evaluated the same during those two years but just want to double check. 
2. I am a little unclear on the expected unit of observation in the Demographics data. We do have a handful (around 15) observations for some year files with duplicates of the studentpseudoid variable. When I check, this is caused by two observations of the same student in a given year with a different schoollocationcode. In  all of these instances the schoollocationcode and preferredlocationcodes for a given observation are the same. What does this mean for our panel in creating observations unique at the studentpseudoid-year level? 
3. In the Suspensions file, data is unique at the studentpseudoid endyear schoollocationcode ofsuspendeddays suspensiontype suspensiondate level. I can in theory reshape the data so it is student-year and then indicators of #suspendeddays, type, date for each observation of this student-year occurrence. But a little uncertain on what it means to have different observations of the same suspension date and type for student-year with different number of suspendeddays. Ideally, I would just sum the # of suspendeddays given for each studentpseudoid-year but I wonder if I might be double counting something.

**Comment by aparthan** [comment]
Another question is regarding the files in the Unified_Enrollment folder. I went through the files and noticed different like mag_name_1, mag_action_1, mag_name_2 and so on for a given student_pseudo_id-school_year-application_id observation. I just want to know if there are only specific measures that we want to merge in to our analysis dataset, or like an indicator for if a given student_pseudo_id-year has an interaction at all for that category. So for example in the analysis dataset we would have a mag_interaction which is 0 or 1 and mag_interaction = 1 if that student-year has an observation in the Deidentified_NEW_MAG_ALL.csv.

**Comment by cqcampos** [comment]
Great -- thanks for the progress. 

Your first three points all sound good. One question about that: how are you managing the work off the server? These files should not leave the server ever. It's ok to create a simulated dataset to test code on off the server but the data should not leave the server. Not saying that you're doing that but just mentioning it in case. 

Some responses to the things you flagged:

1. Yes, a covid issue with the missing years. 
2. The unit of observation in the Demographics files should a student. I imagine this is happening for students that switch schools during the school year. It's strange that two are reported because the demographics files are supposed to contain end-of-year enrollment. If you have any additional information in the data that allows you to see which one is end-of-year, then go with that obs. Otherwise, drop at random but set a seed so the build is the same everytime we run it. 
3. Good catch. I think this means that a student got suspended for several incidents that occurred in a day. It's ok to sum across them.

**Comment by cqcampos** [comment]
About the choice data, let's merge in everything. LA is kind of a mess in how they run their choice programs. MAG_1, MAG_2, etc are schools students applied to. Every choice program is a separate application though and that is why you have so many different files. See the link below for more details

https://echoices.lausd.net/#gsc.tab=0

So for now merge in everything, but rename the variables in the various choice datasets so that we know the program they correspond to. Does that make sense?

**Comment by aparthan** [comment]
Okay thanks for the responses on 1-3. I think I can go ahead and get the Suspensions  and SBA merged with the demographics and update the code so you can see that when you land.

**Comment by aparthan** [comment]
@cqcampos  PFA a document overviewing the current progress with the analysis build. The main point I would like us to discuss today is merging in the Unified_Enrollment data which I have attempted to individually create into student-year form. The main issue I am running into is just stata crashing with the number of variables. 



[Prepare for Data Analysis Task #4.pdf](https://github.com/cqcampos/lausdinfoRCT/files/13861554/Prepare.for.Data.Analysis.Task.4.pdf)

**Comment by cqcampos** [comment]
@aparthan Thanks for the document above. It was very helpful and will inform the meeting with Declan. I sent you a major update through Slack. Respond to any of the important items through here so we maintain a record of the important ones here and not flood this chat with less important communication. Some remaining things:

1. Please finish the build portion of this, incorporating some of my Slack comments, by 1/11 if you can. This gives me the weekend to look over it in detail and make sure we are ready for the next stage. 
2. Ryan finished the geocoding and Declan will send us the relevant data in the next day or so. I can take care of incorporating that into the build when it arrives. 
3. Please add any data inconsistencies/issues you find to the document you have above. We will send this to Declan at some point before we meet with him. 


Thanks for your continued work on this.
